### ISSUE TITLE
Problems with training in multi-gpu caching images

### ISSUE THREAD

**Issue by hdnh2006** [ultralytics/yolov5#2237]
## ❔Question

I am trying to train a large dataset into a "super computer": 488GB RAM, 8 Tesla K80, 32 cores.

The dataset contains more than 200K images.

Everything seems run good while caching images:
![image](https://user-images.githubusercontent.com/17271049/108199729-bc2d2e80-711d-11eb-97dd-4e3a8ed8f73a.png)

But when the process finish, it tries to do it again:
![image](https://user-images.githubusercontent.com/17271049/108199851-e7b01900-711d-11eb-92c5-a49f396dcd7e.png)

What is happening here? Why YOLO tries to make the caching twice. I think 488GB RAM should be enough to train this model, the dataset has 100GB of images.

Should be this enough to train this model? Why YOLO is trying to cache twice the images?

**Comment by glenn-jocher** [comment]
@hdnh2006 you should be very cautious trying to cache 200k images. The dataset size on your hard drive is not directly related to the amount of memory it will consume when cached. COCO is 20GB when stored on disk but 125 GB when cached into RAM (at 640).

Without a full screenshot or terminal printout it's hard to tell what's going on there for you, as you have two seperate screenshots and the relation is unlcear.

**Comment by hdnh2006** [comment]
I see, anyway it is rare YOLO tries to cache the images twice. Thanks Glenn.

**Comment by glenn-jocher** [comment]
@hdnh2006 well it is possible that multi-gpu caching might have an issue, as it's been less used. In our own experience on COCO, caching is just as fast as using 8-12 dataloader workers on a very fast machine (fast CPUs and memory read speeds, i.e. 1-2 GB/s), though for smaller datasets caching also help significantly, and it also helps particularly if the dataset images are large, i.e. 4k images, because they will be read in and resized to --img-size only once.

If you spot any issues or come up with any solutions please let us know!

**Comment by jmj23** [comment]
I seem to have a related experience. 
Here is the output in the console from the cache setup (caching to disk):
```
train: Scanning '/home/cloudservice/data/LLesion/train_image_list' images and labels...87304 found, 0 missing, 4700
train: New cache created: /home/cloudservice/data/LLesion/train_image_list.cache
train: Caching images (181.3GB disk): 100%|█████████████████████████████████| 87304/87304 [07:07<00:00, 204.03it/s]
train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
val: Scanning '/home/cloudservice/data/LLesion/valid_image_list' images and labels...19763 found, 0 missing, 10612 
  0%|                                                                                    | 0/87304 [00:00<?, ?it/s]val: New cache created: /home/cloudservice/data/LLesion/valid_image_list.cache
train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:53<00:00, 1644.18it/s]
train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:53<00:00, 1636.61it/s]
train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:55<00:00, 1581.50it/s]
train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:56<00:00, 1550.32it/s]
train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:56<00:00, 1548.21it/s]
train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:59<00:00, 1466.06it/s]
train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [01:01<00:00, 1427.10it/s]
val: Caching images (41.0GB disk): 100%|████████████████████████████████████| 19763/19763 [01:41<00:00, 195.04it/s]
```

This is running on a GCP VM. It seems that around 300GB of space is used, instead of about 180GB as shown.
nproc_per_node is 8.
Let me know if I can post any other help info for this issue. I'm not completely confident that it's actually a bug, as you say the caching takes up larger than expected space anyways, so I have just been spec'ing out very large boot drives and it seems to work fine. It would be nice to be able to use smaller drives if that were possible though.

**Comment by glenn-jocher** [comment]
@jmj23 it doesn't say 180GB, it shows 222GB used just for caching, not to mention your actual dataset and your OS.

We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by glenn-jocher** [comment]
> I seem to have a related experience. Here is the output in the console from the cache setup (caching to disk):
> 
> ```
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list' images and labels...87304 found, 0 missing, 4700
> train: New cache created: /home/cloudservice/data/LLesion/train_image_list.cache
> train: Caching images (181.3GB disk): 100%|█████████████████████████████████| 87304/87304 [07:07<00:00, 204.03it/s]
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
> train: Scanning '/home/cloudservice/data/LLesion/train_image_list.cache' images and labels... 87304 found, 0 missin
> val: Scanning '/home/cloudservice/data/LLesion/valid_image_list' images and labels...19763 found, 0 missing, 10612 
>   0%|                                                                                    | 0/87304 [00:00<?, ?it/s]val: New cache created: /home/cloudservice/data/LLesion/valid_image_list.cache
> train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:53<00:00, 1644.18it/s]
> train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:53<00:00, 1636.61it/s]
> train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:55<00:00, 1581.50it/s]
> train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:56<00:00, 1550.32it/s]
> train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:56<00:00, 1548.21it/s]
> train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [00:59<00:00, 1466.06it/s]
> train: Caching images (181.3GB disk): 100%|████████████████████████████████| 87304/87304 [01:01<00:00, 1427.10it/s]
> val: Caching images (41.0GB disk): 100%|████████████████████████████████████| 19763/19763 [01:41<00:00, 195.04it/s]
> ```
> 
> This is running on a GCP VM. It seems that around 300GB of space is used, instead of about 180GB as shown. nproc_per_node is 8. Let me know if I can post any other help info for this issue. I'm not completely confident that it's actually a bug, as you say the caching takes up larger than expected space anyways, so I have just been spec'ing out very large boot drives and it seems to work fine. It would be nice to be able to use smaller drives if that were possible though.

In your example you haven't displayed before and after disk usage, i.e. `df -h` screenshots **before you cache** and after you cache.

**Comment by glenn-jocher** [comment]
@jmj23 also note --cache will save images at the --imgsz you specify, so training at higher --imgsz will use more memory.

**Comment by Abesyuffcjk** [comment]
[]() @
Duplicate of #

**Comment by jmj23** [comment]
Fair enough.
I am thinking that even with all this accounted for there is still some extra space being used. This is one of our smaller datasets so the discrepancy is less dramatic. 
I'll post back if I can get more specific accounting of something being off.
Thanks! And thanks for all your work on this fantastic repo.

**Comment by jiegeng321** [comment]
I have a related experience, caching to disk twice. actually another cach is on RAM. my dataset need 1TB in disk cache，i have enough disk space，but when it train，caching again and it is using RAM space，i do not have 1TB RAM,  so it is finished because out of ram memory.  Is it repaired now. @glenn-jocher

**Comment by glenn-jocher** [comment]
@jiegeng321 great! Yes all --cache disk issues should be resolved now for Multi-GPU.

**Comment by hdnh2006** [comment]
I can confirm I have not experimented this issue again, I think Ultralytics team already solve it in the new versions. 

I will close this issue.
